// smooch-core/lib/smooch.js
'use strict';

var _apps = require('./api/apps');

var _menu = require('./api/menu');

var _stripe = require('./api/stripe');

var _appUsers = require('./api/appUsers');

var _disabled = require('./api/disabled');

var _webhooks = require('./api/webhooks');

var _attachments = require('./api/attachments');

var _integrations = require('./api/integrations');

var _deployments = require('./api/deployments');

var _conversations = require('./api/conversations');

var _serviceAccounts = require('./api/serviceAccounts');

var _auth = require('./utils/auth');

var _jwt = require('./utils/jwt');

var jwt = _interopRequireWildcard(_jwt);

var _jsonwebtoken = require('jsonwebtoken');

var _package = require('../package.json');

var _package2 = _interopRequireDefault(_package);

var _templates = require('./api/templates');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var SERVICE_URL = 'https://api.smooch.io';

if (!global.FormData) {
    global.FormData = require('form-data');
}

var Smooch = function Smooch() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, Smooch);

    var _options$serviceUrl = options.serviceUrl,
        serviceUrl = _options$serviceUrl === undefined ? SERVICE_URL : _options$serviceUrl,
        httpAgent = options.httpAgent;

    var auth = {
        keyId: options.keyId,
        secret: options.secret,
        scope: options.scope,
        jwt: options.jwt,
        userId: options.userId
    };

    if (auth.keyId || auth.secret) {
        if (!auth.scope) {
            throw new Error('Invalid auth: missing scope.');
        }

        if (!auth.keyId) {
            throw new Error('Invalid auth: missing keyId.');
        }

        if (!auth.secret) {
            throw new Error('Invalid auth: missing secret.');
        }
    }

    if (auth.keyId && auth.secret && auth.scope) {
        var jwtBody = {
            scope: auth.scope
        };

        if (auth.userId) {
            jwtBody.userId = auth.userId;
        }

        auth.jwt = jwt.generate(jwtBody, auth.secret, auth.keyId);
    } else if (auth.jwt) {
        var decoded = (0, _jsonwebtoken.decode)(auth.jwt);
        if (!decoded) {
            throw new Error('jwt is malformed.');
        }
        if (!decoded.scope) {
            throw new Error('jwt has no scope defined.');
        }
        auth.scope = decoded.scope;
    }

    this.httpAgent = httpAgent;
    this.serviceUrl = serviceUrl;
    this.VERSION = _package2.default.version;
    this.scope = auth.scope;
    this.authHeaders = (0, _auth.getAuthenticationHeaders)(auth);
    this.utils = {};

    this.menu = new _menu.MenuApi(this);
    this.webhooks = new _webhooks.WebhooksApi(this);
    this.attachments = new _attachments.AttachmentsApi(this);
    this.appUsers = new _appUsers.AppUsersApi(this);
    this.conversations = new _conversations.ConversationsApi(this);
    this.stripe = new _stripe.StripeApi(this);
    this.templates = new _templates.TemplatesApi(this);

    if (this.scope === 'account') {
        this.apps = new _apps.AppsApi(this);
        this.integrations = new _integrations.IntegrationsApi(this);
        this.deployments = new _deployments.DeploymentsApi(this);
        this.serviceAccounts = new _serviceAccounts.ServiceAccountsApi(this);
    } else {
        var disabled = new _disabled.DisabledApi('This API requires account level scope');
        this.integrations = this.apps = disabled;
    }

    Object.assign(this.utils, {
        jwt: jwt
    });
};

module.exports = Smooch;
