// smooch-core/lib/api/serviceAccounts.js
'use strict';

exports.__esModule = true;
exports.ServiceAccountsApi = undefined;

var _base = require('./base');

var _serviceAccountKeys = require('./serviceAccountKeys');

var _smoochMethod = require('../utils/smoochMethod');

var _smoochMethod2 = _interopRequireDefault(_smoochMethod);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * @constructor
 * @name ServiceAccountsApi
 * @extends BaseApi
 */
var ServiceAccountsApi = exports.ServiceAccountsApi = function (_BaseApi) {
    _inherits(ServiceAccountsApi, _BaseApi);

    function ServiceAccountsApi() {
        _classCallCheck(this, ServiceAccountsApi);

        var _this = _possibleConstructorReturn(this, _BaseApi.apply(this, arguments));

        _this.requireAppId = false;
        _this.allowedAuth = ['jwt'];

        _this.keys = new (Function.prototype.bind.apply(_serviceAccountKeys.ServiceAccountKeysApi, [null].concat(Array.prototype.slice.call(arguments))))();
        return _this;
    }

    return ServiceAccountsApi;
}(_base.BaseApi);

Object.assign(ServiceAccountsApi.prototype, {
    /**
     * Create a new service account
     * @memberof ServiceAccountsApi.prototype
     * @method create
     * @param  {object} data
     * @return {APIResponse}
     */
    create: (0, _smoochMethod2.default)({
        params: ['name'],
        path: '/serviceaccounts',
        func: function create(url, name) {
            if (typeof name !== 'string') {
                throw new Error('Invalid name parameter type, expected string');
            }

            return this.request('POST', url, {
                name: name
            });
        }
    }),

    /**
     * Fetch the service accounts currently configured
     * @memberof ServiceAccountsApi.prototype
     * @method list
     * @param  {number} limit
     * @param  {number} offset
     * @return {APIResponse}
     */
    list: (0, _smoochMethod2.default)({
        params: ['limit', 'offset'],
        optional: ['limit', 'offset'],
        path: '/serviceaccounts',
        func: function list(url, limit, offset) {
            var q = {};
            if (limit) {
                if (typeof limit !== 'number') {
                    throw new Error('limit must be a number');
                }
                q.limit = limit;
            }
            if (offset) {
                if (typeof offset !== 'number') {
                    throw new Error('offset must be a number');
                }
                q.offset = offset;
            }

            if (q.limit || q.offset) {
                return this.request('GET', url, q);
            } else {
                return this.request('GET', url);
            }
        }
    }),

    /**
     * Retrieve an existing service account
     * @memberof ServiceAccountsApi.prototype
     * @method get
     * @param  {string} serviceAccountId
     * @return {APIResponse}
     */
    get: (0, _smoochMethod2.default)({
        params: ['serviceAccountId'],
        path: '/serviceaccounts/:serviceAccountId',
        method: 'GET'
    }),

    /**
     * Update an existing service account
     * @memberof ServiceAccountsApi.prototype
     * @method put
     * @param  {string} serviceAccountId
     * @return {APIResponse}
     */
    update: (0, _smoochMethod2.default)({
        params: ['serviceAccountId', 'data'],
        path: '/serviceaccounts/:serviceAccountId',
        method: 'PUT'
    }),

    /**
     * Delete an existing service account
     * @memberof ServiceAccountsApi.prototype
     * @method delete
     * @param  {string} serviceAccountId
     * @return {APIResponse}
     */
    delete: (0, _smoochMethod2.default)({
        params: ['serviceAccountId'],
        path: '/serviceaccounts/:serviceAccountId',
        method: 'DELETE'
    })
});
