// smooch-core/lib/api/serviceAccountKeys.js
'use strict';

exports.__esModule = true;
exports.ServiceAccountKeysApi = undefined;

var _base = require('./base');

var _smoochMethod = require('../utils/smoochMethod');

var _smoochMethod2 = _interopRequireDefault(_smoochMethod);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * @constructor
 * @name ServiceAccountKeysApi
 * @extends BaseApi
 */
var ServiceAccountKeysApi = exports.ServiceAccountKeysApi = function (_BaseApi) {
    _inherits(ServiceAccountKeysApi, _BaseApi);

    function ServiceAccountKeysApi() {
        _classCallCheck(this, ServiceAccountKeysApi);

        var _this = _possibleConstructorReturn(this, _BaseApi.apply(this, arguments));

        _this.requireAppId = false;
        _this.allowedAuth = ['jwt'];
        return _this;
    }

    return ServiceAccountKeysApi;
}(_base.BaseApi);

Object.assign(ServiceAccountKeysApi.prototype, {
    /**
     * Create a new service account key
     * @memberof ServiceAccountKeysApi.prototype
     * @method create
     * @param  {string} name
     * @return {APIResponse}
     */
    create: (0, _smoochMethod2.default)({
        params: ['serviceAccountId', 'name'],
        path: '/serviceaccounts/:serviceAccountId/keys',
        func: function create(url, serviceAccountId, name) {
            if (typeof name !== 'string') {
                throw new Error('Invalid name parameter type, expected string');
            }
            return this.request('POST', url, {
                name: name
            });
        }
    }),

    /**
     * Fetch the keys currently configured for a service account
     * @memberof ServiceAccountKeysApi.prototype
     * @method list
     * @return {APIResponse}
     */
    list: (0, _smoochMethod2.default)({
        params: ['serviceAccountId'],
        path: '/serviceaccounts/:serviceAccountId/keys',
        method: 'GET'
    }),

    /**
     * Retrieve an existing service account key
     * @memberof ServiceAccountKeysApi.prototype
     * @method get
     * @param  {string} keyId
     * @return {APIResponse}
     */
    get: (0, _smoochMethod2.default)({
        params: ['serviceAccountId', 'keyId'],
        path: '/serviceaccounts/:serviceAccountId/keys/:keyId',
        method: 'GET'
    }),

    /**
     * Generate a JWT with account scope using the specified key
     * @memberof ServiceAccountKeysApi.prototype
     * @method getJwt
     * @param  {string} keyId
     * @return {APIResponse}
     */
    getJwt: (0, _smoochMethod2.default)({
        params: ['serviceAccountId', 'keyId'],
        path: '/serviceaccounts/:serviceAccountId/keys/:keyId/jwt',
        method: 'GET'
    }),

    /**
     * Delete an existing service account key
     * @memberof ServiceAccountKeysApi.prototype
     * @method delete
     * @param  {string} keyId
     * @return {APIResponse}
     */
    delete: (0, _smoochMethod2.default)({
        params: ['serviceAccountId', 'keyId'],
        path: '/serviceaccounts/:serviceAccountId/keys/:keyId',
        method: 'DELETE'
    })
});
